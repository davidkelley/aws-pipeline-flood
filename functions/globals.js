/**
* Determines the level of logging that the function will output. This enables
* us to reduce the amount of "noise" generated by a battle-tested function,
* whilst enabling us to output additional logging on a new version.
*
* Setting `process.env.LOG_LEVEL` will overwrite the default value of `"info"`
* for this variable.
*
* @type {String}
*/
export const LOG_LEVEL = (process.env.LOG_LEVEL || 'info');

/**
 * The AWS Region that this Lambda function is currently executing in.
 *
 * _Note: This is normally populated by the Lambda execution environment._
 *
 * @type {String}
 */
export const AWS_REGION = process.env.AWS_REGION;


/**
 * The secret API Key used to communicate with the Flood.IO account
 *
 * @type {String}
 */
export const FLOOD_API_KEY = process.env.FLOOD_API_KEY;

/**
 * The URL used to communicate with the Flood.IO API
 *
 * @type {String}
 */
export const FLOOD_API_URI = 'https://api.flood.io/';

/**
 * The path used for sending requests to action Floods
 *
 * @type {String}
 */
export const FLOODS_PATH = 'floods';

/**
 * The path used for sending requests to action Grids
 *
 * @type {String}
 */
export const GRIDS_PATH = 'grids';
