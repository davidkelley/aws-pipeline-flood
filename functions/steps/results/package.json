{
  "name": "results",
  "version": "0.0.0",
  "private": true,
  "main": "./results.js"
}
