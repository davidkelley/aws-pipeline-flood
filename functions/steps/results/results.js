import request from 'request-promise-native';
import { FLOOD_API_KEY, FLOOD_API_URI, FLOODS_PATH } from '../../globals';

/**
 * Retrieves the summary of the flood job.
 *
 * See: https://github.com/flood-io/api-docs/blob/master/endpoints/floods/GET-floods-flood_id-report.md
 *
 * Callback function will be called with an object containing the summary of the flood
 * ```
 * {
 *  summary, string
 * }
 * ```
 *
 * @param {Object} event - object containing the flood uuid
 * @param {Object} context - Lambda runtime information
 * @param {Function} callback - function that will be invoked with (err) or (null, response)
 */
export function handler(
  event = {},
  context,
  callback = () => {}
) {
  const { flood: { uuid } } = event;
  const options = {
    uri: `${FLOOD_API_URI}${FLOODS_PATH}/${uuid}/report`,
    auth: {
      user: FLOOD_API_KEY,
    },
  };

  request.get(options)
    .then(res => {
      const { summary } = JSON.parse(res);
      callback(null, summary);
    })
    .catch(err => {
      callback(err);
    });
}
