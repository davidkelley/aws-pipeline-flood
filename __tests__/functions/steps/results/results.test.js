/* eslint-disable import/no-unresolved */

import request from 'request-promise-native';
import { handler } from '@functions/steps/results/results';
import nock from 'nock';
import { FLOOD_API_URI, FLOODS_PATH } from '@functions/globals';

const mockResponse = {
  summary: 'mock',
};
const mockUUID = '123123';

let requestSpy;
let callbackSpy;

beforeAll(() => {
  nock.disableNetConnect();
  requestSpy = jest.spyOn(request, 'get');
});

afterAll(() => {
  nock.enableNetConnect();
});

describe('Result of flood suite', () => {
  describe('Handles successful query', () => {
    beforeEach((done) => {
      nock(`${FLOOD_API_URI}${FLOODS_PATH}/${mockUUID}`)
        .get('/report')
        .reply(200, mockResponse, { 'Content-Type': 'application/json' });

      callbackSpy = jest.fn().mockImplementation(() => {
        done();
      });

      handler({
        flood: {
          uuid: mockUUID,
        },
      }, null, callbackSpy);
    });

    describe('Calling handler', () => {
      it('Requested the flood results', () => {
        expect(requestSpy.mock.calls).toMatchSnapshot();
      });

      it('Returned the summary to the callback', () => {
        expect(callbackSpy.mock.calls).toMatchSnapshot();
      });
    });
  });

  describe('Handles unsuccesful response ', () => {
    beforeEach((done) => {
      nock(`${FLOOD_API_URI}${FLOODS_PATH}/${mockUUID}`)
        .get('/report')
        .reply(404);

      callbackSpy = jest.fn().mockImplementation(() => {
        done();
      });

      handler({
        flood: {
          uuid: mockUUID,
        },
      }, null, callbackSpy);
    });

    describe('Calling handler', () => {
      it('Returned the error to the callback', () => {
        expect(callbackSpy.mock.calls).toMatchSnapshot();
      });
    });
  });
});
