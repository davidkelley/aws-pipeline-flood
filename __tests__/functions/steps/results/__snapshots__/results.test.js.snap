// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Result of flood suite Handles successful query Calling handler Requested the flood results 1`] = `
Array [
  Array [
    Object {
      "auth": Object {
        "user": "mock-api-key",
      },
      "uri": "https://api.flood.io/floods/123123/report",
    },
  ],
]
`;

exports[`Result of flood suite Handles successful query Calling handler Returned the summary to the callback 1`] = `
Array [
  Array [
    null,
    "mock",
  ],
]
`;

exports[`Result of flood suite Handles unsuccesful response  Calling handler Returned the error to the callback 1`] = `
Array [
  Array [
    [StatusCodeError: 404 - ""],
  ],
]
`;
